src/mmw/apps/geoprocessing_api/calcs.py

```diff
@@ -50,6 +50,70 @@ def animal_population(geojson):
     }
 
 
+def stream_data(geojson):
+    """
+    Given a GeoJSON shape, retreive stream data from the `nhdflowline` table
+    to display in the Analyze tab
+
+    Returns a dictionary to append to outgoing JSON for analysis results.
+    """
+
+    geom = GEOSGeometry(geojson, srid=4326)
+    table_name = 'nhdflowline'
+
+    sql = '''
+        SELECT lengthkm, stream_order
+        FROM {table_name}
+        WHERE ST_Intersects(geom, ST_SetSRID(ST_GeomFromText(%s), 4326))
+        '''.format(table_name=table_name)
+
+    with connection.cursor() as cursor:
+        cursor.execute(sql, [geom.wkt])
+
+        if cursor.rowcount:
+            columns = [col[0] for col in cursor.description]
+            streams = [
+                dict(zip(columns,
+                         [
+                             float(row[0]) if row[0] else 0,
+                             int(row[1]) if row[1] and row[1] != 0 else 999,
+                         ]))
+                for row in cursor.fetchall()
+            ]
+        else:
+            streams = []
+
+    stream_data = {
+        str(x): {
+            "stream_order": x,
+            "lengthkm": 0,
+            "slopepct": 0,
+            "ag_length": 0,
+            "non_ag_length": 0,
+        } for x in list(range(1, 11)) + [999]
+    }
+
+    for s in streams:
+        stream_data[str(s['stream_order'])]['lengthkm'] += s['lengthkm']
+        # TODO: implement `slopepct`, `ag_length`, `non_ag_length`
+
+    stream_data = [
+        {
+            "order": stream_data[k]['stream_order'],
+            "lengthkm": stream_data[k]['lengthkm'],
+            "slopepct": stream_data[k]['slopepct'],
+            "ag_length": stream_data[k]['ag_length'],
+            "non_ag_length": stream_data[k]['non_ag_length'],
+        } for k in stream_data
+    ]
+
+    return {
+        'displayName': 'Streams',
+        'name': 'streams',
+        'categories': sorted(stream_data, key=lambda e: e['order']),
```

Comment:
This is optional and a matter of taste, but I find `itemgetter` easier to read here:

```python
'categories': sorted(stream_data, key=itemgetter('order')),
```

See [here](https://github.com/WikiWatershed/model-my-watershed/blob/20ad7af7a407085b3848d5b474a4562cc86d5d3e/src/mmw/apps/bigcz/clients/cuahsi/search.py#L207) for an example.

Reply:
Sounds good -- I'll make this update.